### PR TITLE
Add access token cnf public key validation in PoP token verifier

### PR DESCRIPTION
**Pull request by AzureMarker**
This check ensures that the user who requested the access token is the same user who created the PoP token.

Also added a few missed error cases and tests to the PoP token verifier.

### CODE REVIEW

**Review by julienstroheker — APPROVED** [review]

File: auth/providers/azure/pop_tokenverifier.go [hunk 1]
```diff
@@ -154,6 +157,23 @@ func (p *PoPTokenVerifier) ValidatePopToken(token string) (string, error) {
...
 
+	if jwk.N == "" || jwk.E == "" {
+		return "", errors.Errorf("Invalid token. 'jwk' claim is empty")
+	}
+
+	// Verify claims from access token
+	var at string
+	if atClaim, ok := claims["at"]; ok {
+		if at, ok = atClaim.(string); !ok {
+			return "", errors.Errorf("Invalid token. at claim should be string")
```
Comment (weinong) [inline auth/providers/azure/pop_tokenverifier.go:168]:
```suggestion
			return "", errors.Errorf("Invalid token. 'at' claim should be string")
```

File: auth/providers/azure/pop_tokenverifier.go [hunk 2]
```diff
@@ -193,6 +213,43 @@ func (p *PoPTokenVerifier) ValidatePopToken(token string) (string, error) {
...
+func (p *PoPTokenVerifier) verifyAccessTokenClaims(publicKey jwk, accessToken string) error {
+	// Parse JWT
+	accessTokenJwt, err := jwt.ParseSigned(accessToken)
+	if err != nil {
+		return fmt.Errorf("could not parse access token in PoP token: %w", err)
+	}
+
+	// Get claims
+	var claims Claims
+	err = accessTokenJwt.UnsafeClaimsWithoutVerification(&claims)
```
Comment (weinong) [inline auth/providers/azure/pop_tokenverifier.go:227]: can you add a comment on why it should be without verification?

File: auth/providers/azure/pop_tokenverifier.go [hunk 3]
```diff
@@ -193,6 +213,43 @@ func (p *PoPTokenVerifier) ValidatePopToken(token string) (string, error) {
...
+	if err != nil {
+		return fmt.Errorf("could not parse claims in access token: %w", err)
+	}
+
+	// Get CNF (confirmation) claim
+	cnf, ok := claims["cnf"]
+	if !ok {
+		return errors.New("could not retrieve cnf claim from access token")
```
Comment (weinong) [inline auth/providers/azure/pop_tokenverifier.go:235]:
```suggestion
		return errors.New("could not retrieve 'cnf' claim from access token")
```

File: auth/providers/azure/pop_tokenverifier.go [hunk 4]
```diff
@@ -193,6 +213,43 @@ func (p *PoPTokenVerifier) ValidatePopToken(token string) (string, error) {
...
+	}
+	cnfMap := map[string]string{}
+	if err := marshalGenericTo(cnf, &cnfMap); err != nil {
+		return fmt.Errorf("failed while parsing cnf in access token: %w", err)
```
Comment (weinong) [inline auth/providers/azure/pop_tokenverifier.go:239]:
```suggestion
		return fmt.Errorf("failed while parsing 'cnf' in access token: %w", err)
```

File: auth/providers/azure/pop_tokenverifier.go [hunk 5]
```diff
@@ -193,6 +213,43 @@ func (p *PoPTokenVerifier) ValidatePopToken(token string) (string, error) {
...
+	}
+
+	// Build the expected public key data
+	jwk := fmt.Sprintf(`{"e":"%s","kty":"RSA","n":"%s"}`, publicKey.E, publicKey.N)
+	jwkS256 := sha256.Sum256([]byte(jwk))
+	encodedJwt := base64.RawURLEncoding.EncodeToString(jwkS256[:])
+
+	if encodedJwt != cnfMap["kid"] {
+		return fmt.Errorf("PoP token validate failed: cnf claim mismatch")
```
Comment (weinong) [inline auth/providers/azure/pop_tokenverifier.go:248]:
```suggestion
		return fmt.Errorf("PoP token validate failed: 'cnf' claim mismatch")
```

**Comment by weinong** [comment]
i actually not quite follow how this change ensures that the user who requested the access token is the same user who created the PoP token

**Comment by AzureMarker** [comment]
> i actually not quite follow how this change ensures that the user who requested the access token is the same user who created the PoP token

The user who created the inner access token gave the PoP token public key to AAD, and the returned access token contains the public key info in the `cnf` claim.

The PoP token wraps the access token, includes the same public key, and is signed by the PoP token private key. The PoP token signature can be verified using the included public key, and the public key in the PoP token claims can be checked against the inner access token's `cnf` claim.

File: auth/providers/azure/pop_tokenverifier.go (hunk 2 above)
Comment (AzureMarker) [reply auth/providers/azure/pop_tokenverifier.go:227]: Updated

**Review by weinong — DISMISSED** [review]
LGTM

**Review by weinong — APPROVED** [review]
LGTM

**Comment by weinong** [comment]
@@tamalsaha I plan to cut a release. Is there any instruction?

**Pull request merged**